torch/_dynamo/variables/tensor.py

```diff
@@ -644,6 +644,46 @@ def redistribute_fn_with_prim_types(x):
                 ),
                 **options,
             )
+        elif name == "register_hook":
+            # see [On tensor.register_hook]
+            assert len(args) == 1
+            fn_var = args[0]
+
+            if isinstance(fn_var, variables.NestedUserFunctionVariable):
+                # NestedUserFunctionVariable don't carry their fn, but reconstruction builds it
+                # This should not be onerous to support when needed.
+                unimplemented("NYI - lambda variables as hooks")
+            elif isinstance(fn_var, variables.functions.FunctoolsPartialVariable):
+                fn = fn_var.as_python_constant()
+                name = fn_var.func.fn.__name__
+            else:
+                fn = fn_var.fn
+                name = fn_var.fn.__name__
+
+            handle = self.as_proxy().node.meta["example_value"].register_hook(fn)
+
+            handle_variable = variables.user_defined.RemovableHandleVariable(
+                handle,
+                user_code_variable_name=None,
+                mutable_local=variables.base.MutableLocal(),
+                **options,
+            )
+            # Note - if the hook has no source, we are forced to lift it up into a global
+            # An example of where this might happen will be if the user creates a custom hook
+            # within a compile region, or wraps an func with a source (like a global)
+            # with a functools partial.  In that case, we can end up in a situation where we
+            # have a source for the internal func, but not the partial. The real fix here is to
+            # not always lift up to globals, but to use the lowest level scope possible to match the
+            # func lifecycle.
```

Comment:
No I don't understand. There's no inner scopes left after you compile. Don't you just need to store it long enough so that you can codegen a register hook for it? Then you can dump it, no global needed.

Reply:
Thats true for inputs, but for intermediaries, we need to keep it alive. This lets us reuse that code. What we can do is lift a weakref tho - that will work.

Reply:
I will change it to also only lift for intermeds

Reply:
Actually, I dont need this.